README.md
# NAME

Distributed::Tasks::Queue - Distributable scalable jobs / tasks processing

# SYNOPSIS

    use strict;
    use Distributed::Tasks::Queue;
    use Plugins::TestOnly;
    use Data::Printer;

    my $jobs_adder  = Distributed::Tasks::Queue->new( );
    my $jobs_worker = Distributed::Tasks::Queue->new( plugin_list => [ Plugins::TestOnly->new ] );

    my $job = {
        id      => 'test_job_one',
        plugin  => 'test_only',
        description => {
            text    => 'To be processed!',
            action  => 'duplicate_text'
        }
    };
    my $res = $jobs_adder->append( $job );
    $jobs_worker->get_jobs( );

and in your plugin, named Plugins::TestOnly

    package Plugins::TestOnly;
    use Moose;
    #ABSTRACT: For testing purposes only. This plugin will transform a text as a job processing example.

    has caller          => ( is => 'rw' );
    has exports_method  => ( is => 'rw', default => sub { return 'test_only' } );

    sub test_only {
      my ( $self, $job ) = @_; 
      warn "PROCESSING JOB...................................................";
      use DDP;
      warn p $job;
      warn "DO WHATEVER............................... the job must be independent and 
                                                       #have every instruction it 
                                                       #needs to be executed";
      if ( $job->{ description }->{ action } eq 'duplicate_text' ) {
        #do whatever.. save on  disk etc
        my $final_text   = $job->{description}->{text} . $job->{description}->{text};
        $job->{ result } = $final_text;
        #save into Database.. etc
        warn $final_text;
        warn "^^ FROM JOB PROCESS";
      }
    }

    sub validate {
      my ( $self, $job ) = @_; 
    }

    1;

# DESCRIPTION

Distributed::Tasks::Queue allows you to queue jobs / tasks that can be processed later. The queue by default is Redis, which could be called as "Data Structure Server". You will need to have it installed and up, unless you use a different queue system.

The distributed tasks queue allows your application to insert a task into a queue. The task must include all the details regarding the task. That will make the task independent and self describing. The task is an $job object (see below). Its usually a hash with instructions like job id, plugin to be used, and the instructions the plugin needs to process the job.

An $job looks like:

    {
      id          => 21876931,                # This should be unique per job
      plugin      => 'image_resizer',         # plugin that will handle this job
      description => {                        # job description.. all the job needs to get done
        you  => 'should make this job independent.',
        that => 'means you should add here all the necessary stuff this job will need',
        for  => 'Example: ',
        save_as => {
            file => '/some/file.png',
            type => '.png'
        },
        options => {
            width     => 200,
            height    => 200,
            auto_crop => 1
        },
        image   => '/some/dir/image.png'
      }
    }

That way you can create a plugin to process each task. Every task must include the plugin name that will handle that task. You should create one plugin for each task. The plugin will receive an object(hash) that you inserted into the queue. That object must have all the information it needs to be processed by your plugin. Your plugin can do whatever... save into a directory, insert into database, etc.

It will use a redis engine by default but you should be able to create a similar backend queue custom class and override the engine. You should be able to override any default atributes also.

I want this module to be generic enough so each user can create custom plugins as they need.

The basic methods are:



append: inserts a the job at the end of queue

prepend: inserts a job at the begining of queue

get\_job\_blocking: gets new jobs in blocking mode. That means if there is nothing on the queue, it will wait untill a lpush or rpush (append/prepend) is executed.

get\_jobs: gets all the jobs from the queue, however its non blocking and will not wait for a job if there is none.

queue\_size: returns the estimated size of the queue

Each of those method will call the respective backend methods, so the queue engine can be anything. By default it uses (Redis::Client).Distributed::Tasks::Queue::Redis

# AUTHOR

Hernan Lopes <hernan@cpan.org>

# COPYRIGHT

Copyright 2013 - Hernan Lopes

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO


Changes
Revision history for Distributed-Tasks-Queue

{{$NEXT}}
        - Initial release


t/lib/Plugins/TestOnly.pm
package Plugins::TestOnly;
use Moose;
#ABSTRACT: For testing purposes only. This plugin will transform a text as a job processing example.

has caller          => ( is => 'rw' );
has exports_method  => ( is => 'rw', default => sub { return 'test_only' } );

sub test_only {
  my ( $self, $job ) = @_; #action: add, list delete , etc
  warn "PROCESSING JOB...................................................";
  use DDP;
  warn p $job;
  warn "DO WHATEVER............................... the job must be independent and have every instruction it needs to be executed";
  if ( $job->{ description }->{ action } eq 'duplicate_text' ) {
    #do whatever.. save on  disk etc
    my $final_text   = $job->{description}->{text}.$job->{description}->{text};
    $job->{ result } = $final_text;
    warn $final_text;
    warn "^^ FROM JOB PROCESS";
  }
}

sub validate {
  my ( $self, $job ) = @_; 
}

1;


lib/Distributed/Tasks/Queue/Redis.pm
package Distributed::Tasks::Queue::Redis;
use Moose;
use JSON::XS;

has queue_name => ( is => 'rw', default => sub { "my_jobs" }  );
has hash_name  => ( is => 'rw', default => sub { "my_hash" }  );
has [ qw/client expiration_time/]     => ( is => 'rw' );

sub insert_on_end {
    my ( $self, $args ) = @_;
    my $count_items = $self->client->rpush(
        $self->queue_name
        , encode_json( $args )
    );
#   $self->client->setex( $args->{ id }, $self->expiration_time, "job id saved for some seconds and will expire" );
    $self->client->hset( $self->hash_name, $args->{ id } => "job id saved for some seconds and will expire" );
}

sub insert_on_begining {
    my ( $self, $args ) = @_;
    my $count_items = $self->client->lpush( 
        $self->queue_name ,
        encode_json( $args )
    );
#   $self->client->setex( $args->{ id }, $self->expiration_time, "job id saved for some seconds and will expire" );
    $self->client->hset( $self->hash_name, $args->{ id } => "job id saved for some seconds and will expire" );
}

sub append {
  my ( $self, $args ) = @_;
  if ( !$self->client->hget( $self->hash_name, $args->{ id } ) ) {
      $self->insert_on_end( $args );
      return 1;
  }
  return 0;
}

sub prepend {
    my ( $self, $args ) = @_; #args = { id => $id, job => $job, expire_date => '...' }
    if ( !$self->client->hget( $self->hash_name, $args->{ id } ) ) {
        $self->insert_on_begining( $args );
        return 1;
    }
    return 0;
}



sub _analyse_job {
    my ( $self, $job ) = @_; 
    if ( defined $job ) {
      my $job = decode_json ( $job );
      if ( exists $job->{ id } ) {
        $self->client->hdel( $self->hash_name, $job->{ id } );
        return $job;
      }
    }
    return 0;
}

=head2 get

process one item from the queue

=cut

sub get {
    my ( $self ) = @_;
    my $job = $self->client->lpop( $self->queue_name );
    return $self->_analyse_job( $job );
}

=head2 get_job_blocking 

process jobs inside a loop, in a blocking way. It will stay in the loop forever waiting for appended or prepended jobs.

=cut

sub get_job_blocking {
  my ( $self ) = @_;
  my ( $queue_name, $job ) = $self->client->blpop( $self->queue_name , 0 ); #TODO can receive a value that can serve as a timeout and can make the process stop... this way a cronjob could restart the jobs processor whenever necessary
  return $self->_analyse_job( $job );
}

=head2 size

Returns the total number of keys stored in the hash. 

The total number of keys should be also the total number of jobs in the queue.

=cut

sub size {
  my ( $self ) = @_; 
  return $self->client->hlen( $self->hash_name );
}


1;


lib/Distributed/Tasks/Queue.pm
package Distributed::Tasks::Queue;
use Moose;
use Distributed::Tasks::Queue::Redis;

use 5.008_005;
our $VERSION = '0.01';
use Redis::Client;

has queue => (
  is        => 'rw', 
  default   => sub {
    return Distributed::Tasks::Queue::Redis->new(
      queue_name        => "my_jobs", 
      hash_name         => "my_hash", 
      expiration_time   => 10,
      client            => Redis::Client->new( host => 'localhost', port => 6379 ),
    )
  }
);

has plugins => (
  is      => "rw",
  default => sub { { } }
);

has plugin_list => (
  is        => 'rw',
  default   => sub { [ qw/
  / ] }
);

sub get_jobs {
    my ( $self ) = @_;
    while ( my $job = $self->queue->get( ) ) {
        $self->process( $job );
    }
}

sub get_job_blocking {
    my ( $self ) = @_;
    while ( my $job = $self->queue->get_job_blocking( ) ) {
        $self->process( $job );
    }
}

sub block_get {
    my ( $self ) = @_; 
    return $self->queue->block_get();
}

sub append {
    my ( $self, $args ) = @_; 
    return $self->queue->append( $args );
#       if $self->can_process( $args->{ plugin } );
}

sub prepend {
    my ( $self, $args ) = @_; 
    return $self->queue->prepend( $args );
#       if $self->can_process( $args->{ plugin } );
}

sub queue_size {
    my ( $self ) = @_; 
    return $self->queue->size();
}

sub process {
    my ( $self, $job ) = @_; 
    if ( $self->can_process( $job->{ plugin } ) ) {
        my $export_name = $job->{ plugin };
        $self->plugins->{ $job->{ plugin } }->$export_name( $job );
    } else {
        warn "Cant process job for plugin: " . $job->{ plugin };
    }
}

sub can_process {
    my ( $self, $plugin ) = @_; 
    return exists $self->plugins->{ $plugin };
}

sub BUILD {
    my ( $self ) = @_;
    map {
        $_->caller( $self );
        $self->plugins->{ $_->exports_method } = $_
            if defined $_->exports_method
               and ! exists $self->plugins->{ $_->exports_method }
        } @{ $self->plugin_list };
}


1;
__END__

=encoding utf-8

=head1 NAME

Distributed::Tasks::Queue - Distributable scalable jobs / tasks processing

=head1 SYNOPSIS

  use strict;
  use Distributed::Tasks::Queue;
  use Plugins::TestOnly;
  use Data::Printer;

  my $jobs_adder  = Distributed::Tasks::Queue->new( );
  my $jobs_worker = Distributed::Tasks::Queue->new( plugin_list => [ Plugins::TestOnly->new ] );

  my $job = {
      id      => 'test_job_one',
      plugin  => 'test_only',
      description => {
          text    => 'To be processed!',
          action  => 'duplicate_text'
      }
  };
  my $res = $jobs_adder->append( $job );
  $jobs_worker->get_jobs( );

and in your plugin, named Plugins::TestOnly

  package Plugins::TestOnly;
  use Moose;
  #ABSTRACT: For testing purposes only. This plugin will transform a text as a job processing example.

  has caller          => ( is => 'rw' );
  has exports_method  => ( is => 'rw', default => sub { return 'test_only' } );

  sub test_only {
    my ( $self, $job ) = @_; 
    warn "PROCESSING JOB...................................................";
    use DDP;
    warn p $job;
    warn "DO WHATEVER............................... the job must be independent and 
                                                     #have every instruction it 
                                                     #needs to be executed";
    if ( $job->{ description }->{ action } eq 'duplicate_text' ) {
      #do whatever.. save on  disk etc
      my $final_text   = $job->{description}->{text} . $job->{description}->{text};
      $job->{ result } = $final_text;
      #save into Database.. etc
      warn $final_text;
      warn "^^ FROM JOB PROCESS";
    }
  }

  sub validate {
    my ( $self, $job ) = @_; 
  }

  1;

=head1 DESCRIPTION

Distributed::Tasks::Queue allows you to queue jobs / tasks that can be processed later. The queue by default is Redis, which could be called as "Data Structure Server". You will need to have it installed and up, unless you use a different queue system.

The distributed tasks queue allows your application to insert a task into a queue. The task must include all the details regarding the task. That will make the task independent and self describing. The task is an $job object (see below). Its usually a hash with instructions like job id, plugin to be used, and the instructions the plugin needs to process the job.

An $job looks like:

  {
    id          => 21876931,                # This should be unique per job
    plugin      => 'image_resizer',         # plugin that will handle this job
    description => {                        # job description.. all the job needs to get done
      you  => 'should make this job independent.',
      that => 'means you should add here all the necessary stuff this job will need',
      for  => 'Example: ',
      save_as => {
          file => '/some/file.png',
          type => '.png'
      },
      options => {
          width     => 200,
          height    => 200,
          auto_crop => 1
      },
      image   => '/some/dir/image.png'
    }
  }

That way you can create a plugin to process each task. Every task must include the plugin name that will handle that task. You should create one plugin for each task. The plugin will receive an object(hash) that you inserted into the queue. That object must have all the information it needs to be processed by your plugin. Your plugin can do whatever... save into a directory, insert into database, etc.

It will use a redis engine by default but you should be able to create a similar backend queue custom class and override the engine. You should be able to override any default atributes also.

I want this module to be generic enough so each user can create custom plugins as they need.

The basic methods are:


append: inserts a the job at the end of queue

prepend: inserts a job at the begining of queue

get_job_blocking: gets new jobs in blocking mode. That means if there is nothing on the queue, it will wait untill a lpush or rpush (append/prepend) is executed.

get_jobs: gets all the jobs from the queue, however its non blocking and will not wait for a job if there is none.

queue_size: returns the estimated size of the queue

Each of those method will call the respective backend methods, so the queue engine can be anything. By default it uses (Redis::Client).Distributed::Tasks::Queue::Redis

=head1 AUTHOR

Hernan Lopes E<lt>hernan@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2013 - Hernan Lopes

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
